dm/pkg/checker/table_structure.go

```diff
@@ -267,114 +311,207 @@ func (c *TablesChecker) checkTableOption(opt *ast.TableOption) *incompatibilityO
+			r.Extra = fmt.Sprintf("fail to get parser for instance %s on sharding %s", instance, c.targetTableID)
+			if len(errCh) == 0 {
+				errCh <- err
+			}
+			return
+		}
+		r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.targetTableID)
+		for _, table := range tables {
+			statement, err := dbutil.GetCreateTableSQL(ctx, db, table.Schema, table.Name)
+			if err != nil {
+				// continue if table was deleted when checking
+				if isMySQLError(err, mysql.ErrNoSuchTable) {
+					continue
 				}
-				stmt, err := parser2.ParseOneStmt(statement, "", "")
-				if err != nil {
-					markCheckError(r, errors.Annotatef(err, "statement %s", statement))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+				if len(errCh) == 0 {
+					errCh <- err
 				}
+				return
+			}
 
-				ctStmt, ok := stmt.(*ast.CreateTableStmt)
-				if !ok {
-					markCheckError(r, errors.Errorf("Expect CreateTableStmt but got %T", stmt))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+			info, err := dbutil.GetTableInfoBySQL(statement, parser2)
+			if err != nil {
+				if len(errCh) == 0 {
+					errCh <- err
 				}
-
-				if c.checkAutoIncrementPrimaryKey {
-					passed := c.checkAutoIncrementKey(instance, schema, table, ctStmt, info, r)
-					if !passed {
-						return r
-					}
+				return
+			}
+			stmt, err := parser2.ParseOneStmt(statement, "", "")
+			if err != nil {
+				if len(errCh) == 0 {
+					errCh <- errors.Annotatef(err, "statement %s", statement)
 				}
+				return
+			}
 
-				if stmtNode == nil {
-					stmtNode = ctStmt
-					firstTable = dbutil.TableName(schema, table)
-					firstInstance = instance
-					continue
+			ctStmt, ok := stmt.(*ast.CreateTableStmt)
+			if !ok {
+				if len(errCh) == 0 {
+					errCh <- errors.Errorf("Expect CreateTableStmt but got %T", stmt)
 				}
+				return
+			}
 
-				checkErr := c.checkConsistency(stmtNode, ctStmt, firstTable, dbutil.TableName(schema, table), firstInstance, instance)
-				if checkErr != nil {
-					r.State = StateFailure
-					r.Errors = append(r.Errors, checkErr)
-					r.Extra = fmt.Sprintf("error on sharding %s", c.name)
-					r.Instruction = "please set same table structure for sharding tables"
-					return r
+			if c.checkAutoIncrementPrimaryKey {
+				passed := c.checkAutoIncrementKey(instance, table.Schema, table.Name, ctStmt, info, r)
+				if !passed {
+					return
 				}
 			}
+
+			checkErr := c.checkConsistency(stmtNode, ctStmt, firstTable, table.String(), firstInstance, instance)
+			if checkErr != nil {
+				r.State = StateFailure
+				r.Errors = append(r.Errors, checkErr)
+				r.Extra = fmt.Sprintf("error on sharding %s", c.targetTableID)
+				r.Instruction = "please set same table structure for sharding tables"
+				return
+			}
+		}
+	}
+
+	for instance, tables := range c.tableMap {
+		concurrency := getConcurrency(len(tables))
+		maxConnections, err := utils.GetMaxConnections(ctx, c.dbs[instance])
```

Comment:
Still not recommend using max connection. Also if there are some other clients hold the connection?

Reply:
This is indeed a problem, but this is just to prevent the upstream from reporting errors **to a certain extent**. Because we also can't get the real number of upstream connections.

Reply:
yes, we should not use whole max connection, use `32` as the max concurrencies is enough when `32` < server max connection due to your test report

i say `adjust concurrent by server's max connection` means we should not try open a connection pool that hold connection count larger than server allow (this may get an error from server, not tested)